Djangofixes

sum12 opened the pull request:
fixed to be able to use https://github.com/sum12/rqlite-django backend

sum12 committed:
Row can now handle slices to __getitme__

slice is another common way to retriev data. So we can rely on tuple to
handle slices for us

sum12 committed:
some ORMs use %s as query formatters regex them to question-marks

since it is using regex there are no false positives for replacements

sum12 committed:
implemented fetchmany by loop over fetchone

here it expects to return as many row as possible.
ret_size = min(max(avail, 1), requested)

sum12 committed:
do not default to using transactions

sum12 committed:
make transactions optional, to not break backward compaitbility

by passing an options to connections we can pass it to allow cursors
based on that connection

otoolep commented:
@zmedico -- any comment?

zmedico commented on src/pyrqlite/cursors.py:
```diff
@@ -148,6 +150,10 @@ def _substitute_params(self, operation, parameters):
         return sql_str.split(None, 1)[0].upper()
 
+    def convert_query(self, query):
+        FORMAT_QMARK_REGEX = re.compile(r'(?<!%)%s')
```
This isn't safe, the parameter style really needs to be adjustable in the ORM, for example sqlalchemy's SQLiteDialect declares `default_paramstyle = 'qmark'` here:

https://github.com/zzzeek/sqlalchemy/blob/rel_1_2_10/lib/sqlalchemy/dialects/sqlite/base.py#L1043

zmedico commented on src/pyrqlite/cursors.py:
```diff
@@ -156,11 +162,14 @@ def execute(self, operation, parameters=None):
         operation = self._substitute_params(operation, parameters)
 
         command = self._get_sql_command(operation)
+        transact = ''
```
When automatic transactions are disabled, it probably makes sense to support the `isolation_level` parameter that python's sqlite3 module supports:

https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.isolation_level

The default behavior, `isolation_level=''`, means to implicitly issue a plain BEGIN statement before a DML statement. The parameter can be set to DEFERRED, IMMEDIATE, or EXCLUSIVE in order to make it implicitly issue the corresponding type of BEGIN statement before a DML statement. For `isolation_level=None`, autocommit mode is triggered, as described here:

https://www.sqlite.org/c3ref/get_autocommit.html

sum12 committed:
added 'default_paramstyle' option to Connection

as of now only two types of param style will be permitted, either
question mark (qmark) or one expected by rqlite (percent). Connection
will default to  qmark style which will be regex replaced in the cursor
object

zmedico commented on src/pyrqlite/connections.py:
```diff
@@ -43,7 +43,7 @@ class Connection(object):
                  user=None, password=None, connect_timeout=None,
                  detect_types=0, max_redirects=UNLIMITED_REDIRECTS,
-                 always_transactions=True):
+                 always_transactions=True, default_paramstyle='qmark'):
```
In my remark about default_paramstyle in sqlalchemy's SQLiteDialect, my intention was to suggest that the ORM conform to the paramstyle of the database. The existing convention is for paramstyle to be constant for a particular database type, see:

https://www.python.org/dev/peps/pep-0249/#paramstyle

sum12 commented:
so iiuc I can drop this change of param style, because django uses `qmark` style  for sqlite ? and that is already handled

and for the `isolation_level` iiuc, django would prefer `isolation_level=''` since it handles transactions by itself, but `pyrqlite` will have to do the right thing for other values of `isolation_level`. Did I get that right ?

zmedico commented:
> so iiuc I can drop this change of param style, because django uses qmark style for sqlite ? and that is already handled

Yes, if django respects the `paramstyle` value defined in src/pyrqlite/dbapi2.py then it should behave correctly without any paramstyle changes in pyrqlite.

> and for the isolation_level iiuc, django would prefer isolation_level='' since it handles transactions by itself, but pyrqlite will have to do the right thing for other values of isolation_level. Did I get that right ?

If django works correctly with the default `isolation_level=''` setting of the sqlite3 module, then it should work fine with pyrqlite too if we implement the `isolation_level` parameter.

sum12 committed:
use passed factory to create a cursor for conneciton

the sqlite module in python does not cache the factory object so every
call is either expected to pass a factory object or get an object of the
default cursor

sum12 committed:
Revert "added 'default_paramstyle' option to Connection"

This reverts commit 365c94e9c0dabc588c22178a0e585b34cef5d094.

sum12 committed:
Revert "make transactions optional, to not break backward compaitbility"

This reverts commit f325fa7fe85f19d8f6e6addca9275668d62ab3de.

sum12 committed:
Revert "do not default to using transactions"

This reverts commit 29c1edf07efbdfe0396fba3eabd53535cb445711.

sum12 committed:
Revert "some ORMs use %s as query formatters regex them to question-marks"

This reverts commit 93519becbb09dd4a81620b05322299e438b9cb23.

sum12 commented:
I have reverted most of my changes. Thanks to @zmedico for the patiently explaining stuff. The changes related to transaction and `isolation_level` are not required after turning off a "feature flag" `autocommits_when_autocommit_is_off` in the sqlite3 backend of django

edit: @zmedico if the changes look good to you, then let me know I will squash them according

zmedico commented on src/pyrqlite/cursors.py:
```diff
@@ -277,8 +277,12 @@ def fetchone(self):
-            size = self.arraysize
-        raise NotImplementedError(self)
+            size = max(self.arraysize, len(self._rows))
+        size = min(len(self._rows), size)
```
The size calculation looks like in could be incorrect for some corner cases. Also, the [fetchmany documentation](https://www.python.org/dev/peps/pep-0249/#fetchmany) says to use arraysize when size has not been specified, so I'd like to conform to that. How about something like this:
```python
    def fetchmany(self, size=None):
        remaining = self.arraysize if size is None else size
        remaining = min(remaining, self.rowcount - self.rownumber)
        for i in range(remaining):
            rows.append(self.fetchone())
        return rows
```

zmedico replied:
I've merged the Row slice change in aa562e4942531e96a13b927b69be4bdf9ed6cc43, and the cursor factory change in 9da1eb12c66e73b425b70523984517ec744d3a5a.

zmedico commented:
Changes merged in aa562e4942531e96a13b927b69be4bdf9ed6cc43, 9da1eb12c66e73b425b70523984517ec744d3a5a, and f8cd41e0522af36a234088b3c876ccfc5078f662. Thanks!
